# AMATH 352B

## Homework Assignment #6 - due 5/20/2022

This Jupyter Notebook consists of several cells.  Some of the cells are **Markdown** cells, and some of the cells are **Code** cells.

Once you have honed your Notebook to the point where you're ready to turn it in, convert the Notebook to HTML and upload both the Notebook (the .ipynb file) and the HTML version of it to Canvas.  There are potentially issues in the conversion involving figures, graphics, and imported files to solve, so leave yourself enough time to get this done.

In [1]:
from math import *
import numpy as np
from scipy.linalg import lu_factor, lu_solve
from scipy.linalg import qr, solve_triangular
import matplotlib as mpl
import matplotlib.pyplot as plt

## Problem #1

By hand, compute the eigenvalues of the matrix

$$\begin{pmatrix}0 & 1 & 0\cr 1 & 0 & 1\cr 0 & 1 & 0\cr\end{pmatrix}$$

$$\lambda = 0, -\sqrt{2}, \sqrt{2}$$

## Problem #2

By hand, compute the eigenvectors corresponding to the eigenvalues you computed in Problem #1

$$
\text{Eigenvectors} =
\begin{pmatrix}-1\cr0\cr1\cr\end{pmatrix},
\begin{pmatrix}1\cr-\sqrt{2}\cr1\cr\end{pmatrix},
\begin{pmatrix}1\cr\sqrt{2}\cr1\cr\end{pmatrix}$$

## Problem #3

Compute the eigenvalues and eigenvectors of that matrix in Python by calling ``np.linalg.eig``.  Compare with your answers to the first two problems

In [2]:
A = np.array([[0, 1, 0], [1, 0, 1], [0, 1, 0]])
evals, evecs = np.linalg.eig(A)
print("Eigenvalues:", evals)
print("Eigenvectors:",evecs)

Eigenvalues: [-1.41421356e+00  9.24824453e-17  1.41421356e+00]
Eigenvectors: [[ 5.00000000e-01  7.07106781e-01  5.00000000e-01]
 [-7.07106781e-01  1.52655666e-16  7.07106781e-01]
 [ 5.00000000e-01 -7.07106781e-01  5.00000000e-01]]


## Problem #4

Where does Gerschgorin's Theorem predict that the eigenvalues of the following matrix will lie?

$$\begin{pmatrix}3 & 1 & 0 & 0\cr
1 & 7 & 1 & 0\cr
0 & 1 & 7 & 1\cr
0 & 0 & 1 & 12\cr\end{pmatrix}$$

$$\text{Gerschgorins Theorem tells us that the eigenvalues will be between:}$$
$$2\leq\lambda\leq4$$
$$5\leq\lambda\leq9$$
$$11\leq\lambda\leq13$$

## Problem #5

Suppose $V$ is the matrix whose columns are given by the unit eigenvectors of the matrix

$$\begin{pmatrix}1 & 0 & 3 & 4\cr
0 & 5 & 2 & 1\cr
3 & 2 & 17 & 0\cr
4 & 1 & 0 & -4\cr\end{pmatrix}$$

What can you say about the matrix $V^TV$?

$$V^TV\text{ is equivalent to } V^2 \text{because } V = V^T$$

## Problem #6

If all eigenvalues $\lambda$ of a matrix $A$ satisfy $|\lambda|<1$, what can be said about the limit of $A^nx$ as $n\to\infty$, regardless of what $x$ is?

$$A = x\lambda x^{-1}$$
$$A^2 = x\lambda x^{-1} * x\lambda x^{-1}$$
$$A^2 = x\lambda \lambda x^{-1}$$

$$\text{The matrix to the power of n will be x multiplied by the matrix of lambdas to the power of n, multiplied by x inverse.}$$
$$\text{The matrix of eigenvalues being multiplied by itself n times with eigenvalues of }|\lambda|<1\text{ means they will continually get smaller and closer to zero}$$

$$\lim\limits_{n\to \infty} (A^n x)$$ 
$$\text{will converge to zero because values } <1 \text{ multiplied over and over converge to zero}$$

## Problem #7

Nancy has been engaged in the effort to save Loggerhead Turtles from extinction for many years.
She is currently tasked with preparing a report for the National Marine Fisheries Service, and she needs your help as an expert in linear algebra to help her analyze the data for her report.  Nancy is able to provide you with a **Leslie matrix** model for the Loggerhead Turtle population

$$L=\begin{pmatrix}0 & 0 & 0 & 0 & 127 & 4 & 80\cr
0.6747x & 0.737 & 0 & 0 & 0 & 0 & 0\cr
0 & 0.0486 & 0.6610 & 0 & 0 & 0 & 0\cr
0 & 0 & 0.0147 & 0.6907y & 0 & 0 & 0\cr
0 & 0 & 0 & 0.0518y & 0 & 0 & 0\cr
0 & 0 & 0 & 0 & 0.8091y & 0 & 0\cr
0 & 0 & 0 & 0 & 0 & 0.8091y & 0.8089y\cr\end{pmatrix}$$

The variables $x$ and $y$ control three different strategies for stabilizing the population of Loggerhead Turtles.  Nancy is preparing a report on the effectiveness of these three strategies for the National Marine Fisheries Service

* **Do nothing**.  Set $x=1$ and $y=1$.  Provide Nancy with the data she needs to show that this strategy leads to extinction of Loggerhead Turtles.
* **Protect eggs and hatchlings**.  Set $0.6747x=1$ and $y=1$.  Provide Nancy with evidence that this strategy will prove ineffective and that extinction of Loggerhead Turtles will still occur even if all the eggs and hatchlings are protected.
* **Equip fishing nets with turtle excluding devices**.  Set $x=1$ and $y=1.15$.  Demonstrate that this measure will prevent the extinction of Loggerhead Turtles.

As usual, your instructor has lots of advice which may or may not be relevant to this problem:

* The National Marine Fisheries Service was founded in 1871
* If the vector $p_n$ represents the number of turtles alive in several age groups at year $n$, then $Lp_n$ represents the number of turtles alive in those same age groups the following year
* Charles throws back all Loggerhead Turtles that he catches from the dock
* Eigenvalues of magnitude less than one can help determine how quickly certain quantities decrease
* Nigel occasionally spots Loggerhead Turtles when he is traveling at 6 knots or less
* Nancy has been a dues paying member of Greenpeace for several years
* Ocean currents affect the migration patterns of sea turtles
* Cubic models of data can be constructed by computing the $QR$ factorization of a **Vandermonde** matrix


In [28]:
# Create leslie matrix with given X and Y
def createLeslie(x, y):
    return np.array([[0, 0, 0, 0, 127, 4, 80],
                   [0.6747*x, 0.737, 0, 0, 0, 0, 0],
                   [0, 0.0486, 0.6610, 0, 0, 0, 0],
                   [0, 0, 0.0147, 0.6907*y, 0, 0, 0],
                   [0, 0, 0, 0.0518*y, 0, 0, 0],
                   [0, 0, 0, 0, 0.8091*y, 0, 0],
                   [0, 0, 0, 0, 0, 0.8091*y, 0.8089*y]])

# "Do nothing" test
evs = np.linalg.eig(createLeslie(1, 1))
print("Do nothing")
for ev in evs[0]:
    final = (abs(ev))
print(final)
# "Protect eggs and hatchlings" test
evs = np.linalg.eig(createLeslie(1/0.6747, 1))
print("\nProtect eggs and hatchlings")
for ev in evs[0]:
    final = (abs(ev))
print(final)
# "Equip fishing nets with turtle excluding devices" test
evs = np.linalg.eig(createLeslie(1, 1.15))
print("\nEquip fishing nets with turtle excluding devices")

for ev in evs[0]:
    final = (abs(ev))
print(final)

Do nothing
0.9450309806910039

Protect eggs and hatchlings
0.9649561636204429

Equip fishing nets with turtle excluding devices
1.0228726429196717


$\text{The best option for the survival/growth of the turtle population is going to be to equip fishing nets with turtle excluding devices since}$
$\text{the survival coefficient for this treatment group will be 1.0228726429196717
which actually grows the population.}$